// rch-driver-js/dist/esm/printer/ReceiptStatus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReceiptStatus = void 0;
/**
 * Receipt status
 * @date 1/11/2022 - 12:38:03 PM
 *
 * @export
 * @enum {number}
 */
var ReceiptStatus;
(function (ReceiptStatus) {
    /**
     * No receipt or document
     * @date 1/11/2022 - 12:38:10 PM
     */
    ReceiptStatus[ReceiptStatus["IDLE"] = 0] = "IDLE";
    /**
     * A fiscal document has started
     * @date 1/11/2022 - 12:38:18 PM
     */
    ReceiptStatus[ReceiptStatus["FISCAL_TRANSITION"] = 1] = "FISCAL_TRANSITION";
    /**
     * The payment phase has started
     * @date 1/11/2022 - 12:38:29 PM
     */
    ReceiptStatus[ReceiptStatus["FISCAL_PAYMENT"] = 2] = "FISCAL_PAYMENT";
    /**
     * Aplha input
     * @date 1/11/2022 - 12:38:38 PM
     */
    ReceiptStatus[ReceiptStatus["ALPHA_INPUT"] = 3] = "ALPHA_INPUT";
    /**
     * Waiting a temrination command
     * @date 1/11/2022 - 12:38:45 PM
     */
    ReceiptStatus[ReceiptStatus["WAITING_TERMINATION"] = 4] = "WAITING_TERMINATION";
})(ReceiptStatus = exports.ReceiptStatus || (exports.ReceiptStatus = {}));
