// rch-driver-js/dist/esm/AbstractDevice.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractDevice = void 0;
const DeviceType_1 = require("./DeviceType");
/**
 * @inheritdoc
 * @date 1/11/2022 - 3:13:37 PM
 *
 * @export
 * @abstract
 * @class AbstractDevice
 * @typedef {AbstractDevice}
 * @implements {IAbstractDevice}
 */
class AbstractDevice {
    constructor() {
        /**
         * @inheritdoc
         */
        this.serialNumber = '';
        /**
         * @inheritdoc
         */
        this.partNumber = '';
        /**
         * @inheritdoc
         */
        this.type = DeviceType_1.DeviceType.PRINTF;
        /**
         * @inheritdoc
         */
        this.ip = '';
        /**
         * @inheritdoc
         */
        this.ipPort = 23;
        /**
         * @inheritdoc
         */
        this.comPort = '';
        /**
         * @inheritdoc
         */
        this.baudRate = 9600;
        /**
         * @inheritdoc
         */
        this.hostName = '';
        /**
         * @inheritdoc
         */
        this.macAddress = '';
        /**
         * @inheritdoc
         */
        this.active = true;
    }
}
exports.AbstractDevice = AbstractDevice;
