// rch-driver-js/dist/esm/EcrDevice.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EcrDevice = void 0;
const AbstractDevice_1 = require("./AbstractDevice");
/**
 * ECR Devoce
 * @inheritdoc
 *
 * @export
 * @class EcrDevice
 * @typedef {EcrDevice}
 * @extends {AbstractDevice}
 * @implements {IEcrDevice}
 */
class EcrDevice extends AbstractDevice_1.AbstractDevice {
    constructor() {
        super(...arguments);
        /**
         * @inheritdoc
         */
        this.fwVersion = '';
        /**
         * @inheritdoc
         */
        this.fwVersionLabel = '';
        /**
         * @inheritdoc
         */
        this.hasProgDump = false;
        /**
         * @inheritdoc
         */
        this.nDepartments = 0;
        /**
         * @inheritdoc
         */
        this.nOperators = 0;
        /**
         * @inheritdoc
         */
        this.nPayments = 0;
        /**
         * @inheritdoc
         */
        this.nVats = 0;
        /**
         * @inheritdoc
         */
        this.prog = null;
        /**
         * @inheritdoc
         */
        this.hasDgfeFreeSpace = false;
    }
}
exports.EcrDevice = EcrDevice;
