// rch-driver-js-core/dist/esm/dto/BillDTO.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BillDTO = void 0;
const BillType_1 = require("./BillType");
/**
 * Interface of a Bill including Receipts, Returns and Cancel Documents
 * @date 1/11/2022 - 10:10:12 AM
 *
 * @export
 * @class BillDTO
 * @typedef {BillDTO}
 */
class BillDTO {
    constructor() {
        /**
         * Bill Type to identify the Bill
         * @date 1/11/2022 - 10:11:59 AM
         *
         * @type {BillType}
         */
        this.billType = BillType_1.BillType.RECEIPT;
        /**
         * The items to be sold or return
         * @date 1/11/2022 - 10:15:33 AM
         *
         * @type {LineItemDTO[]}
         */
        this.lineItems = [];
    }
    /**
     * This is true if the BillType is a RETURN
     * @date 1/11/2022 - 10:12:52 AM
     *
     * @readonly
     * @type {boolean}
     */
    get isReturn() {
        return this.billType == BillType_1.BillType.RETURN;
    }
    /**
     * This is true if the BillType is a CANCEL
     * @date 1/11/2022 - 10:14:08 AM
     *
     * @readonly
     * @type {boolean}
     */
    get isCancel() {
        return this.billType == BillType_1.BillType.CANCEL;
    }
    /**
     * This is true if the BillType is a RECEIPT
     * @date 1/11/2022 - 10:14:45 AM
     *
     * @readonly
     * @type {boolean}
     */
    get isReceipt() {
        return this.billType == BillType_1.BillType.RECEIPT;
    }
}
exports.BillDTO = BillDTO;
