// rch-driver-js/dist/esm/models/Department.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Department = void 0;
const AbstractParser_1 = require("./AbstractParser");
/**
 * @inheritdoc
 *
 * @export
 * @class Department
 * @typedef {Department}
 * @extends {AbstractParser}
 * @implements {IDepartment}
 */
class Department extends AbstractParser_1.AbstractParser {
    /**
     * Creates an instance of Department.
     * @date 1/11/2022 - 12:12:44 PM
     *
     * @constructor
     * @param {string} firstRow
     * @param {string} secondRow
     */
    constructor(firstRow, secondRow) {
        super();
        this.firstRow = firstRow;
        this.secondRow = secondRow;
        this.id = parseInt(firstRow.substring(1, 4));
        this.name = firstRow.substring(4, 24);
        this.name.trim();
        this.price = this.parseIntWithDecimal(firstRow.substring(24, 34));
        this.halo = this.parseIntWithDecimal(secondRow.substring(4, 14));
        this.lalo = this.parseIntWithDecimal(secondRow.substring(14, 24));
        this.vatCode = this.indexChartToInt(secondRow.charCodeAt(28));
        this.single = secondRow.charAt(25) == '1';
        this.groupCode = parseInt(secondRow.substring(26, 28));
        this.departmentType = this.indexChartToInt(secondRow.charCodeAt(28));
    }
}
exports.Department = Department;
